<a href="https://colab.research.google.com/github/Ashahet1/AI-Engineer-Roadmap-2024/blob/main/Chapter_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install tensorflow-datasets

In [ ]:
import tensorflow as tf
import tensorflow_datasets as tfds
mnist_train = tfds.load(name="fashion_mnist", split="train")
assert isinstance(mnist_train, tf.data.Dataset)
print(type(mnist_train))

Dl Completed...: 0 url [00:00, ? url/s]

Dl Size...: 0 MiB [00:00, ? MiB/s]

Extraction completed...: 0 file [00:00, ? file/s]

Generating splits...:   0%|          | 0/2 [00:00<?, ? splits/s]

Generating train examples...:   0%|          | 0/60000 [00:00<?, ? examples/s]

Shuffling /root/tensorflow_datasets/fashion_mnist/incomplete.IF5ZL6_3.0.1/fashion_mnist-train.tfrecord*...:   …

Generating test examples...:   0%|          | 0/10000 [00:00<?, ? examples/s]

Shuffling /root/tensorflow_datasets/fashion_mnist/incomplete.IF5ZL6_3.0.1/fashion_mnist-test.tfrecord*...:   0…

Dataset fashion_mnist downloaded and prepared to /root/tensorflow_datasets/fashion_mnist/3.0.1. Subsequent calls will reuse this data.
<class 'tensorflow.python.data.ops.prefetch_op._PrefetchDataset'>


In [ ]:
for item in mnist_train.take(1):
  print(type(item))
  print(item.keys())
  print(item['image'])
  print(item['label'])

mnist_test, info = tfds.load(name="fashion_mnist", with_info="true")
print(info)

<class 'dict'>
dict_keys(['image', 'label'])
tf.Tensor(
[[[  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [ 18]
  [ 77]
  [227]
  [227]
  [208]
  [210]
  [225]
  [216]
  [ 85]
  [ 32]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]]

 [[  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [ 61]
  [100]
  [ 97]
  [ 80]
  [ 57]
  [117]
  [227]
  [238]
  [115]
  [ 49]
  [ 78]
  [106]
  [108]
  [ 71]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]]

 [[  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [ 81]
  [105]
  [ 80]
  [ 69]
  [ 72]
  [ 64]
  [ 44]
  [ 21]
  [ 13]
  [ 44]
  [ 69]
  [ 75]
  [ 75]
  [ 80]
  [114]
  [ 80]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]]

 [[  0]
  [  0]
  [  0]
  [  0]
  [  0]
  [ 26]
  [ 92]
  [ 69]
  [ 68]
  [ 75]
  [ 75]
  [ 71]
  [ 74]
  [ 83]
  [ 75]
  [ 77]
  [ 78]
  [ 74]
  [ 74]
  [ 83]
  [ 77]
  [108]
  [ 34]
  [  0]
  [  0]
  [  0]
  [  0]
  [  0]]

 [[  0]
  [  0]
  [  0]
  [  0]
  [  0]


**TFDS with Keras Models**
Example 1: Fashion to learn by model types of shoes

In [ ]:
import tensorflow as tf
import tensorflow_datasets as tfds

(training_images, training_labels), (test_images, test_labels) =  tfds.as_numpy(tfds.load('fashion_mnist', split = ['train', 'test'], batch_size=-1, as_supervised=True))

training_images = training_images / 255.0
test_images = test_images / 255.0

model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28,1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs=5)

/usr/local/lib/python3.10/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.7626 - loss: 0.6760
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.8555 - loss: 0.4065
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8635 - loss: 0.3714
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.8743 - loss: 0.3457
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step - accuracy: 0.8780 - loss: 0.3300


Example 2 more Complex instance on Horses or Human Binary Classification using convolution neural network layer.

In [ ]:
import tensorflow as tf
import tensorflow_datasets as tfds

data = tfds.load('horses_or_humans', split='train', as_supervised=True)

train_batches = data.shuffle(100).batch(10)

model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(16, (3,3), activation='relu',
                           input_shape=(300, 300, 3)),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

model.compile(optimizer='Adam', loss='binary_crossentropy',
metrics=['accuracy'])

history = model.fit(train_batches, epochs=10)

Dl Completed...: 0 url [00:00, ? url/s]

Dl Size...: 0 MiB [00:00, ? MiB/s]

Generating splits...:   0%|          | 0/2 [00:00<?, ? splits/s]

Generating train examples...:   0%|          | 0/1027 [00:00<?, ? examples/s]

Shuffling /root/tensorflow_datasets/horses_or_humans/incomplete.8Z3KWQ_3.0.0/horses_or_humans-train.tfrecord*.…

Generating test examples...:   0%|          | 0/256 [00:00<?, ? examples/s]

Shuffling /root/tensorflow_datasets/horses_or_humans/incomplete.8Z3KWQ_3.0.0/horses_or_humans-test.tfrecord*..…

Dataset horses_or_humans downloaded and prepared to /root/tensorflow_datasets/horses_or_humans/3.0.0. Subsequent calls will reuse this data.
Epoch 1/10


/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


103/103 ━━━━━━━━━━━━━━━━━━━━ 97s 901ms/step - accuracy: 0.7293 - loss: 7.0723
Epoch 2/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 154s 1s/step - accuracy: 0.9569 - loss: 0.1301
Epoch 3/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 125s 867ms/step - accuracy: 0.9802 - loss: 0.0716
Epoch 4/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 92s 895ms/step - accuracy: 0.9697 - loss: 0.0987
Epoch 5/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 92s 898ms/step - accuracy: 0.9832 - loss: 0.0523
Epoch 6/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 141s 886ms/step - accuracy: 0.9836 - loss: 0.0345
Epoch 7/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 100s 966ms/step - accuracy: 0.9970 - loss: 0.0113
Epoch 8/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 131s 864ms/step - accuracy: 0.9248 - loss: 0.2991
Epoch 9/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 145s 891ms/step - accuracy: 0.9593 - loss: 0.1135
Epoch 10/10
103/103 ━━━━━━━━━━━━━━━━━━━━ 139s 857ms/step - accuracy: 0.9940 - loss: 0.0257


In above example Accuracy is achieved of about 99.40% on the train dataset.

In [ ]:
val_data = tfds.load('horses_or_humans', split='test', as_supervised=True)
validation_batches = val_data.batch(32)
history = model.fit(train_batches, epochs = 8, validation_data = validation_batches, validation_steps = 1)

Epoch 1/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 97s 938ms/step - accuracy: 0.9995 - loss: 0.0017 - val_accuracy: 0.8125 - val_loss: 1.7012
Epoch 2/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 134s 867ms/step - accuracy: 0.9912 - loss: 0.0314 - val_accuracy: 0.8125 - val_loss: 4.2984
Epoch 3/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 88s 858ms/step - accuracy: 0.9713 - loss: 0.0915 - val_accuracy: 0.8438 - val_loss: 9.0723
Epoch 4/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 145s 892ms/step - accuracy: 0.9775 - loss: 0.0896 - val_accuracy: 0.9688 - val_loss: 0.2202
Epoch 5/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 88s 858ms/step - accuracy: 0.9952 - loss: 0.0196 - val_accuracy: 0.9375 - val_loss: 1.6302
Epoch 6/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 142s 856ms/step - accuracy: 1.0000 - loss: 4.9399e-04 - val_accuracy: 0.9062 - val_loss: 1.5846
Epoch 7/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 140s 840ms/step - accuracy: 1.0000 - loss: 3.3000e-04 - val_accuracy: 0.7500 - val_loss: 11.2449
Epoch 8/8
103/103 ━━━━━━━━━━━━━━━━━━━━ 89s 863ms/step - accuracy: 1.0000 - lo

**ETL (EXTRACT, TRANSFORM AND LOAD)**

ETL is the core pattern that TensorFlow uses for training, regardless of scale. We’ve been exploring small-scale, single-computer model building in this book, but the same technology can be used for large-scale training across multiple machines with massive datasets.

The *Extract phase* of the ETL process is when the raw data is loaded from wherever it is stored and prepared in a way that can be transformed. The *Transform phase* is when the data is manipulated in a way that makes it suitable or improved for training. For example, batching, image augmentation, mapping to feature columns, and other such logic applied to the data can be considered part of this phase. The *Load phase* is when the data is loaded into the neural network for training.

In [ ]:
# Source from https://stackoverflow.com/questions/61380552/unable-to-import-tensorflow-addons
!pip install tensorflow-addons==0.19.0


In [ ]:
!pip uninstall tensorflow-addons
!pip uninstall tensorflow

Found existing installation: tensorflow-addons 0.16.1
Uninstalling tensorflow-addons-0.16.1:
  Would remove:
    /usr/local/lib/python3.10/dist-packages/_foo.cpython-310-x86_64-linux-gnu.so
    /usr/local/lib/python3.10/dist-packages/tensorflow_addons-0.16.1.dist-info/*
    /usr/local/lib/python3.10/dist-packages/tensorflow_addons/*
Proceed (Y/n)? Y
  Successfully uninstalled tensorflow-addons-0.16.1
Found existing installation: tensorflow 2.18.0
Uninstalling tensorflow-2.18.0:
  Would remove:
    /usr/local/bin/import_pb_to_tensorboard
    /usr/local/bin/saved_model_cli
    /usr/local/bin/tensorboard
    /usr/local/bin/tf_upgrade_v2
    /usr/local/bin/tflite_convert
    /usr/local/bin/toco
    /usr/local/bin/toco_from_protos
    /usr/local/lib/python3.10/dist-packages/tensorflow-2.18.0.dist-info/*
    /usr/local/lib/python3.10/dist-packages/tensorflow/*
Proceed (Y/n)? Y
  Successfully uninstalled tensorflow-2.18.0


In [ ]:
!pip install tensorflow==2.18.0

  Using cached tensorflow-2.18.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.1 kB)
Using cached tensorflow-2.18.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (615.3 MB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tf-keras 2.17.0 requires tensorflow<2.18,>=2.17, but you have tensorflow 2.18.0 which is incompatible.


There was an error I was struggling to import tensoflow-addons installed different versions But using below command which would use command with square brackets ([]) and "tensorflow" as the extra dependency. This is the standard syntax that PIP follows to search and find the correct versions that are compatible. The previous command included compatible_with, which isn't the standard practice to use with pip.

In [ ]:
!pip install "tensorflow-addons[tensorflow]"

  Using cached tensorflow_addons-0.23.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.8 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 475.2/475.2 MB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 611.8/611.8 kB 35.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 60.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 56.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 60.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 442.0/442.0 kB 25.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 5.8 MB/s eta 0:00:00
  Attempting uninstall: wrapt
    Found existing installation: wrapt 1.16.0
    Uninstalling wrapt-1.16.0:
      Successfully uninstalled wrapt-1.16.0
  Attempting uninstall: typeguard
    Found existing installation: typeguard 4.4.1
    Uninstalling typeguard-4.4.1:
      Successfully uninstalled typeguard-4.4.1
  Attempting unin

In [ ]:
import tensorflow_addons as tfa
import tensorflow as tf
import tensorflow_datasets as tfds

# MODEL DEFINITION START #
model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(16, (3,3), activation='relu',
                           input_shape=(300, 300, 3)),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])
model.compile(optimizer='Adam', loss='binary_crossentropy',
              metrics=['accuracy'])
# MODEL DEFINITION END #

# EXTRACT PHASE START #
data = tfds.load('horses_or_humans', split='train', as_supervised=True)
val_data = tfds.load('horses_or_humans', split='test', as_supervised=True)
# EXTRACT PHASE END

# TRANSFORM PHASE START #
def augmentimages(image, label):
  image = tf.cast(image, tf.float32)
  image = (image/255)
  image = tf.image.random_flip_left_right(image)
  image = tfa.image.rotate(image, 40, interpolation='NEAREST') # tfa is used for rotate, flip, ship etc.
  return image, label

train = data.map(augmentimages)
train_batches = train.shuffle(100).batch(32)
validation_batches = val_data.batch(32)
# TRANSFORM PHASE END

# LOAD PHASE START #
history = model.fit(train_batches, epochs=5,
                    validation_data=validation_batches, validation_steps=1)
# LOAD PHASE END #

/usr/local/lib/python3.10/dist-packages/tensorflow_addons/utils/tfa_eol_msg.py:23: UserWarning: 

TensorFlow Addons (TFA) has ended development and introduction of new features.
TFA has entered a minimal maintenance and release mode until a planned end of life in May 2024.
Please modify downstream libraries to take dependencies from other repositories in our TensorFlow community (e.g. Keras, Keras-CV, and Keras-NLP). 

For more information see: https://github.com/tensorflow/addons/issues/2807 

  warnings.warn(


Epoch 1/5
33/33 [==============================] - 119s 3s/step - loss: 0.4798 - accuracy: 0.7653 - val_loss: 186.7708 - val_accuracy: 0.6875
Epoch 2/5
33/33 [==============================] - 103s 3s/step - loss: 0.2196 - accuracy: 0.9192 - val_loss: 223.5135 - val_accuracy: 0.7188
Epoch 3/5
33/33 [==============================] - 96s 3s/step - loss: 0.1706 - accuracy: 0.9328 - val_loss: 356.3870 - val_accuracy: 0.5938
Epoch 4/5
33/33 [==============================] - 102s 3s/step - loss: 0.0627 - accuracy: 0.9757 - val_loss: 449.4225 - val_accuracy: 0.5625
Epoch 5/5
33/33 [==============================] - 95s 3s/step - loss: 0.0605 - accuracy: 0.9737 - val_loss: 326.2792 - val_accuracy: 0.7500


# Parallelizing ETL to Improve Training Performance

The cache method will cache the dataset in memory. If you have a lot of RAM available this is a really useful speedup. Trying this in Colab with Dogs vs. Cats will likely crash your VM due to the dataset not fitting in RAM. After that, if available, the Colab infrastructure will give you a new, higher-RAM machine.

In [12]:
# Create a model architecture
import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflow_addons as tfa
import numpy as np
import multiprocessing

def create_model():
  model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(16, (3,3), activation='relu',
                                      input_shape=(300, 300, 3)),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
  ])


  model.compile(optimizer='Adam', loss='binary_crossentropy', metrics=['accuracy'])
  return model

In [20]:
# Extract
train_data = tfds.load('cats_vs_dogs', split='train', with_info=True)
data_dir = info.data_dir

# Construct the file pattern using the data directory
file_pattern = f'{data_dir}/cats_vs_dogs-train.tfrecord*'

# Print the file pattern to check if it's correct
print(f"Looking for files with pattern: {file_pattern}")

# List the files matching the pattern
files = tf.data.Dataset.list_files(file_pattern)

train_dataset = files.interleave(
                     tf.data.TFRecordDataset,
                     cycle_length=4,
                     num_parallel_calls=tf.data.experimental.AUTOTUNE
                )



Looking for files with pattern: /root/tensorflow_datasets/cats_vs_dogs/4.0.1/cats_vs_dogs-train.tfrecord*


In [21]:
# Transform
def read_tfrecord(serialized_example):
  feature_description={
      "image": tf.io.FixedLenFeature((), tf.string, ""),
      "label": tf.io.FixedLenFeature((), tf.int64, -1),
  }
  example = tf.io.parse_single_example(
       serialized_example, feature_description
  )
  image = tf.io.decode_jpeg(example['image'], channels=3)
  image = tf.cast(image, tf.float32)
  image = image / 255
  image = tf.image.resize(image, (300,300))
  return image, example['label']

In [22]:
cores = multiprocessing.cpu_count()
print(cores)

train_dataset = train_dataset.map(
    read_tfrecord, num_parallel_calls=cores
)
# train_dataset = train_dataset.cache()

2


Yes, it crashed
The cache method will cache the dataset in memory. If you have a lot of RAM available this is a really useful speedup. Trying this in Colab with Dogs vs. Cats will likely crashed my VM due to the dataset not fitting in RAM. After that, if available, the Colab infrastructure will give a new, higher-RAM machine. But for me it never asked any pop etc for more RAM memory so I don't think i might be able to run again.

In [ ]:
# Load

train_dataset = train_dataset.shuffle(1024).batch(32) # Use the same batch size as before

train_dataset = train_dataset.prefetch(tf.data.experimental.AUTOTUNE) # Use the same prefetch size as before

model = create_model()

model.fit(train_dataset, epochs=8, verbose=1)


Epoch 1/8
     10/Unknown - 37s 3s/step - loss: 0.7001 - accuracy: 0.5094